### PR TITLE
feat: Add support for geometry collection parsing from WKT to GeoJSON

### PR DESCRIPTION
**Pull request by Zakhmaster**
Added the ability to convert WKT Geometry Collection to a GeoJSON Geometry Collection. Required simple changes to the Yacc file and a few helper functions. 

Added tests for most subsets of possible geometries in the collection
Added tests for empty collection
Added tests for Z and/or M properties
Added test to cover scenario of a collection inside of a collection

### CODE REVIEW

**Comment by Zakhmaster** [comment]
Seems like when I ran `npm run release` it changed line endings in a bunch of files causing the diff to show many changes. I will try to fix that.

**Comment by jgravois** [comment]
very cool contribution @Zakhmaster. i should be able to review it more fully within a week or so.

> Seems like when I ran `npm run release` it changed line endings in a bunch of files causing the diff to show many changes. I will try to fix that.

you're not going to be able to publish these changes to npm. generally i'm a proponent of bumping version numbers and prepping for a release _after_ PRs with new features or bug fixes are merged.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,11 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
```
Comment (jgravois) [inline CHANGELOG.md:6]: this is the place to reference changes that have yet to be published...

Comment (Zakhmaster) [reply CHANGELOG.md:6]: Aha, I see. Sorry I deal with changelogs often but have not had to deal with a "unreleased" change yet.

**Comment by jgravois** [comment]
thx again for the PR and tweaks @Zakhmaster! i'd usually be able to help land something like this quickly, but i have a few too many plates in the air at the moment. 🤹 

in the meantime, thanks for being patient.

**Comment by Zakhmaster** [comment]
> thx again for the PR and tweaks @Zakhmaster! i'd usually be able to help land something like this quickly, but i have a few too many plates in the air at the moment. 🤹
> 
> in the meantime, thanks for being patient.

Hey, no need to rush. It is not critical for me.

**Comment by Zakhmaster** [comment]
I am not sure what to do about the merge checks. If there is something I can do to get those passing let me know.

File: packages/wkt/test/wkt.test.js [hunk 2]
```diff
@@ -604,3 +604,404 @@ test('should parse a MULTIPOLYGON with a hole', function (t) {
...
+    properties: {
+      z: true
+    }
+  });
+});
+
+test('should parse a GEOMETRYCOLLECTION with M property', function (t) {
+  t.plan(1);
+
+  const input = 'GEOMETRYCOLLECTION M ( POINT (40 10), LINESTRING (10 10, 20 20, 10 40), POLYGON ((40 40, 20 45, 45 30, 40 40)))';
```
Comment (jgravois) [inline packages/wkt/test/wkt.test.js:866]: you're calling this a geometry collection with M values, but none of the coordinates have a third value...

File: packages/wkt/test/wkt.test.js [hunk 3]
```diff
@@ -604,3 +604,404 @@ test('should parse a MULTIPOLYGON with a hole', function (t) {
...
+        ]
+      }
+    ]
+  });
+});
+
+test('should parse a GEOMETRYCOLLECTION with ZM property', function (t) {
+  t.plan(1);
+
+  const input = 'GEOMETRYCOLLECTION ZM ( POINT (40 10), LINESTRING (10 10, 20 20, 10 40), POLYGON ((40 40, 20 45, 45 30, 40 40)))';
```
Comment (jgravois) [inline packages/wkt/test/wkt.test.js:971]: same here. neither Z nor M values are actually present in the geometries.

according to the [GeoJSON RFC](https://www.rfc-editor.org/rfc/rfc7946#section-3.1.1), Z values are supported but M values are not, so ideally this parser would preserve the first and omit the second.

**Review by jgravois — CHANGES_REQUESTED** [review]
thanks again for taking the time to put together this contribution and write tests. 🌟 

i confirmed they are indeed passing locally. not sure whats up with CI/CD and can't be bothered to go down that wormhole tonight. 😅

File: packages/wkt/test/wkt.test.js (hunk 2 above)
Comment (Zakhmaster) [reply packages/wkt/test/wkt.test.js:866]: I struggled to find a standard for this scenario. It was unclear if Specifying M or Z on GeometryCollection directly forces it's children to have a third attribute. You probably know the correct resolution here. It seems entirely probable that a collection could support both 2D, 3D and 4D attributes. To me the Z and M feel like noop values.

Comment (Zakhmaster) [reply packages/wkt/test/wkt.test.js:866]: I think it's fair to assume that if geometry collection has one of the attributes, then all of it's children likely would as well. I'll go forward with that assumption and fix the tests.

File: packages/wkt/test/wkt.test.js (hunk 3 above)
Comment (Zakhmaster) [reply packages/wkt/test/wkt.test.js:971]: I could achieve this by simply removing the fourth insert on line 56 of wkt.yy. But that sounds like breaking changes outside the scope of the PR. I am simply using existing parsing to conveniently handle collection parsing.

Comment (jgravois) [reply packages/wkt/test/wkt.test.js:971]: > I could achieve this by simply removing the fourth insert on line 56 of wkt.yy

i see. that's something i would welcome in a PR of its own.

File: packages/wkt/test/wkt.test.js (hunk 2 above)
Comment (Zakhmaster) [reply packages/wkt/test/wkt.test.js]: I hope these changes reflect what I intended above

File: packages/wkt/src/wkt.yy [hunk 4]
```diff
@@ -42,6 +43,8 @@ expressions
         { return $1; }
     | multipolygon EOF
         { return $1; }
+    | geometrycollection EOF
+        { return $1;}
```
Comment (jgravois) [inline packages/wkt/src/wkt.yy:47]: just a nit:
```suggestion
        { return $1; }
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,6 +3,12 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+
+## [2.2.0]
+### Added
+* `@terraformer/wkt`
+  * Parse WKT Geometry Collection to GeoJSON Geomtery Collection.
+
 ## [Unreleased]
```
[the remark this answers is not in the document]
Comment (jgravois) [reply CHANGELOG.md:11]: i'm still waiting on you to incorporate my suggested change above ☝️

File: packages/wkt/test/wkt.test.js [hunk 6]
```diff
@@ -604,3 +604,434 @@ test('should parse a MULTIPOLYGON with a hole', function (t) {
...
+        ]
+      }
+    ]
+  });
+});
+
+test('should parse a GEOMETRYCOLLECTION with ZM property', function (t) {
+  t.plan(1);
+
+  const input = 'GEOMETRYCOLLECTION ZM ( POINT ZM (40 10 10 15), LINESTRING ZM (10 10 10 15, 20 20 10 15, 10 40 10 15), POLYGON ZM ((40 40 10 15, 20 45 10 15, 45 30 10 15, 40 40 10 15)))';
```
Comment (jgravois) [inline packages/wkt/test/wkt.test.js:989]: this is much better!
PRs that include comprehensive tests are a maintainers dream ☁️🌈

**Review by jgravois — COMMENTED** [review]
this is _extremely_ close to being ready to merge.

File: CHANGELOG.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zakhmaster) [reply CHANGELOG.md:11]: My bad, I miss-read it. working on it today.

[the remark this answers is not in the document]
Comment (Zakhmaster) [reply CHANGELOG.md]: Finally have time today to make such a simple change.

**Comment by Zakhmaster** [comment]
@jgravois I think I have addressed all of the comments? Let me know!

File: CHANGELOG.md [hunk 7]
```diff
@@ -3,8 +3,14 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+
+
 ## [Unreleased]
```
Comment (jgravois) [inline CHANGELOG.md:8]: i'm going to push a commit of my own to remove the extra whitespace in your branch.

```suggestion
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
```

**Review by jgravois — APPROVED** [review]
nice work. thanks again! 💪

**Pull request merged**